File: c++/Cpu.cpp
#include <iostream>
using namespace std;
#include "Komponen.cpp"
#include "Overclockable.h"
#include "Compatible.h"

class Cpu : public Komponen, public Overclockable, public Compatible {
private:
    double kecepatanGHz;
    int jumlahCore;
    string socket;

public:
    Cpu(string merk, int tahunProduksi, double harga, string kode, string tipe, double kecepatanGHz, int jumlahCore, string socket);
    double getKecepatanGHz();
    void setKecepatanGHz(double kecepatanGHz);
    int getJumlahCore();
    void setJumlahCore(int jumlahCore);
    string getSocket();
    void setSocket(string socket);
    double calculatePerformance();
    void displayInfo() override;
    void overclock(double increase) override;
    void resetToDefault() override;
    bool isOverclocked() override;
    bool checkCompatibility(Komponen* komponen) override;
    string* getCompatibleWith() override;
};

double Cpu::getKecepatanGHz() { return kecepatanGHz; }
void Cpu::setKecepatanGHz(double kecepatanGHz) { this->kecepatanGHz = kecepatanGHz; }

int Cpu::getJumlahCore() { return jumlahCore; }
void Cpu::setJumlahCore(int jumlahCore) { this->jumlahCore = jumlahCore; }

string Cpu::getSocket() { return socket; }
void Cpu::setSocket(string socket) { this->socket = socket; }

double Cpu::calculatePerformance() {
    return kecepatanGHz * jumlahCore;
}

void Cpu::displayInfo() {
    Komponen::displayInfo();
    cout << "Kecepatan GHz: " << kecepatanGHz << endl;
    cout << "Jumlah Core: " << jumlahCore << endl;
    cout << "Socket: " << socket << endl;
}

void Cpu::overclock(double increase) {
    this->kecepatanGHz += increase;
}

void Cpu::resetToDefault() {
    this->kecepatanGHz = 3.0; // Default value
}

bool Cpu::isOverclocked() {
    return this->kecepatanGHz > 3.0;
}

bool Cpu::checkCompatibility(Komponen* komponen) {
    // Dummy implementation
    return true;
}

string* Cpu::getCompatibleWith() {
    // Dummy implementation
    static string compatibleWith[] = {"Intel", "AMD"};
    return compatibleWith;
}

File: java/Cpu.java
public class Cpu extends Komponen implements Overclockable, Compatible {
    private double kecepatanGHz;
    private int jumlahCore;
    private String socket;

    public Cpu(String merk, int tahunProduksi, double harga, String kode, String tipe, double kecepatanGHz, int jumlahCore, String socket) {
        super(merk, tahunProduksi, harga, kode, tipe);
        this.kecepatanGHz = kecepatanGHz;
        this.jumlahCore = jumlahCore;
        this.socket = socket;
    }

    public double getKecepatanGHz() { return kecepatanGHz; }
    public void setKecepatanGHz(double kecepatanGHz) { this.kecepatanGHz = kecepatanGHz; }

    public int getJumlahCore() { return jumlahCore; }
    public void setJumlahCore(int jumlahCore) { this.jumlahCore = jumlahCore; }

    public String getSocket() { return socket; }
    public void setSocket(String socket) { this.socket = socket; }

    public double calculatePerformance() {
        return kecepatanGHz * jumlahCore;
    }

    @Override
    public void displayInfo() {
        super.displayInfo();
        System.out.println("Kecepatan GHz: " + kecepatanGHz);
        System.out.println("Jumlah Core: " + jumlahCore);
        System.out.println("Socket: " + socket);
    }

    public void overclock(double increase) {
        this.kecepatanGHz += increase;
    }

    public void resetToDefault() {
        this.kecepatanGHz = 3.0; // Default value
    }

    public boolean isOverclocked() {
        return this.kecepatanGHz > 3.0;
    }

    public boolean checkCompatibility(Komponen komponen) {
        // Dummy implementation
        return true;
    }

    public String[] getCompatibleWith() {
        // Dummy implementation
        return new String[]{"Intel", "AMD"};
    }
}